src/linalg.jl

```diff
@@ -188,6 +188,22 @@ function _A_mul_Bt_or_Bc!(tfun::Function, C::StridedMatrix, A::AbstractMatrix, B
     C
 end
 
+function *(A::Diagonal, b::AbstractSparseVector)
+    if size(A, 2) != length(b)
+        throw(
+            DimensionMismatch("The dimension of the matrix A $(size(A)) and of the vector b $(length(b))")
+        )
+    end
+    T = promote_eltype(A, b)
+    res = similar(b, T)
+    nzind_b = nonzeroinds(b)
+    nzval_b = nonzeros(b)
+    for idx in eachindex(nzind_b)
+        res[nzind_b[idx]] = A.diag[nzind_b[idx]] * nzval_b[idx]
```

Comment:
I don't get the change, what is `y` here? We are not multiplying in-place no?

Reply:
The point is that writing into a sparse vector by indexing is very slow, because it needs to do all sorts of index checks (for instance, check if it has to add a new nonzero entry), all of which are unnecessary here, because the `nonzeroinds` remain exactly the same.

Reply:
Just to demonstrate:

```julia
using LinearAlgebra, BenchmarkTools, SparseArrays

d = randn(1000);
D = Diagonal(d);
x = sprandn(1000, 0.2);
@btime $D*$x; # 4.338 μs (22 allocations: 72.25 KiB)

import Base: *
using Base: promote_eltype
using SparseArrays: nonzeroinds, nonzeros

function *(A::Diagonal, b::AbstractSparseVector)
    if size(A, 2) != length(b)
        throw(
            DimensionMismatch(lazy"The dimension of the matrix A $(size(A)) and of the vector b $(length(b))")
        )
    end
    T = promote_eltype(A, b)
    res = similar(b, T)
    nzind_b = nonzeroinds(b)
    nzval_b = nonzeros(b)
    for idx in eachindex(nzind_b)
        res[nzind_b[idx]] = A.diag[nzind_b[idx]] * nzval_b[idx]
    end
    return res
end

@btime $D*$x; #  2.416 μs (4 allocations: 3.59 KiB)

function *(A::Diagonal, b::AbstractSparseVector)
    if size(A, 2) != length(b)
        throw(
            DimensionMismatch(lazy"The dimension of the matrix A $(size(A)) and of the vector b $(length(b))")
        )
    end
    T = promote_eltype(A, b)
    res = similar(b, T)
    nzind_b = nonzeroinds(b)
    nzval_b = nonzeros(b)
    nzval_y = nonzeros(res)
    for idx in eachindex(nzind_b)
        nzval_y[idx] = A.diag[nzind_b[idx]] * nzval_b[idx]
    end
    return res
end

@btime $D*$x; # 214.257 ns (4 allocations: 3.59 KiB)
```